src/serializer/functions.js

```diff
@@ -57,26 +61,40 @@ export class Functions {
   writeEffects: WriteEffects;
   _noopFunction: void | ECMAScriptSourceFunctionValue;
 
+  _unwrapAbstract(value: AbstractValue): Value {
+    let elements = value.values.getElements();
+    if (elements) {
+      let possibleValues = [...elements].filter(
+        element => !(element instanceof EmptyValue || element instanceof UndefinedValue)
+      );
+      if (possibleValues.length === 1) {
+        return possibleValues[0];
+      }
+    }
+    return value;
+  }
+
   __optimizedFunctionEntryOfValue(value: Value): AdditionalFunctionEntry | void {
     let realm = this.realm;
-
+    // if we conditionally called __optimize, we may have an AbstractValue that is the union of Empty or Undefined and
+    // a function/component to optimize
     if (value instanceof AbstractValue) {
-      // if we conditionally called __optimize, we may have an AbstractValue that is the union of Empty or Undefined and
-      // a function/component to optimize
-      let elements = value.values.getElements();
-      if (elements) {
-        let possibleValues = [...elements].filter(
-          element => !(element instanceof EmptyValue || element instanceof UndefinedValue)
-        );
-        if (possibleValues.length === 1) {
-          value = possibleValues[0];
-        }
-      }
+      value = this._unwrapAbstract(value);
     }
-    if (value instanceof ECMAScriptSourceFunctionValue) {
-      // additional function logic
-      return { value };
-    } else if (value instanceof ObjectValue) {
+    invariant(value instanceof ObjectValue);
+    let funcValue = Get(realm, value, "funcValue");
+    if (funcValue !== realm.intrinsics.undefined) {
+      // unwrap from this side
+      if (funcValue instanceof AbstractValue) {
+        funcValue = this._unwrapAbstract(funcValue);
+      }
+      // regular case with __optimized with type information
+      invariant(funcValue instanceof ECMAScriptSourceFunctionValue);
+      let argModelString = Get(realm, value, "argModelString");
+      return argModelString instanceof StringValue
```

Comment:
If you had to `_unwrapAbstract` the `funcValue` above, then you probably also have to unwrap the `argModelString` here.

Reply:
It is not possible because a non-string argument to `__optimize` will produce an error. Should I support it? It is unlikely that models are going to be provided as something dynamically evaluated.

Reply:
I now understand why you did what you did. It's okay.